File: 56-merge-intervals/56-merge-intervals.java
class Solution {
    public int[][] merge(int[][] intervals) {
        Arrays.sort(intervals, new Comparator<int[]>() {
            @Override
            public int compare(int[] o1, int[] o2) {
                if (o1[0]==o2[0]) return o1[1]-o2[1];
                return o1[0]-o2[0];
            }
        });
        
        int length = intervals.length;
        List<int[]> res = new ArrayList<int[]>();
        
        int left;
        int right;
        //boolean restart = true;
        
        if (length <= 1) return intervals;
        
        //初始化
        left = intervals[0][0];
        right = intervals[0][1];
        
        for (int i = 1; i < length; i++) {
            if (intervals[i][0] > right) {  //与上一区间不重合
                int[] temp = new int[2];
                temp[0] = left;
                temp[1] = right;
                res.add(temp);
                
                left = intervals[i][0];
                right = intervals[i][1];
                
            } else {    //与上一区间重合
                left = left < intervals[i][0] ? left : intervals[i][0];
                right = right > intervals[i][1] ? right : intervals[i][1];
            }
            
            if (i == length - 1) {  //当前为最后一个区间
                int[] temp = new int[2];
                temp[0] = left;
                temp[1] = right;
                res.add(temp);
            }
        }
        
        return res.toArray(new int[res.size()][2]); //记住 list.size() 用法
    }
}

File: 56-merge-intervals/56-merge-intervals.cpp
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        sort(intervals.begin(), intervals.end());
        
        vector<vector<int>> res;
        vector<int> cur(2);
        cur = intervals[0];
        
        int pos = 1;
        while (pos < intervals.size()) {
            if (intervals[pos][0] > cur[1]) { // no overlap
                res.push_back(cur);
                cur = intervals[pos];
            } else {
                cur[0] = min(cur[0], intervals[pos][0]);
                cur[1] = max(cur[1], intervals[pos][1]);
            }
            
            pos++;
        }
        res.push_back(cur);
        
        
        return res;
    }
};